astropy/units/format/ogip.py

```diff
@@ -205,34 +205,31 @@ def p_unit_expression(p):
                             | UNIT OPEN_PAREN complete_expression CLOSE_PAREN power numeric_power
                             | OPEN_PAREN complete_expression CLOSE_PAREN power numeric_power
             """
-            # If we run p[1] in cls._functions, it will try and parse each
-            # item in the list into a unit, which is slow. Since we know that
-            # all the items in the list are strings, we can simply convert
-            # p[1] to a string instead.
-            p1_str = str(p[1])
```

Comment:
Just to be clear - the comment was correct in saying that all items in the list are strings, but the list it was referring to was `cls._functions`, not `p`.

Reply:
Yes, but it is a problem only if `p[1]` can actually be a unit, which I don't think is true. Or is it?

Reply:
`p[1]` is not allowed to be a unit, but if the input is invalid then it might be, and in that case it will trigger a `ValueError` with the `bad_multiplication_message`.

If `cls._functions` is a `frozenset` then `p[1] in cls._functions` is fast both if `p[1]` is a string or a unit.